projects/netgrif-components-core/src/lib/groups/services/ldap-group-list.service.ts

```diff
@@ -0,0 +1,250 @@
+
+
+    /**
+     * Inject services.
+     * Initialize declared attributes.
+     *
+     * Loading and mapped stream of ldap groups.
+     * @param _resources Loading ldap groups from backend.
+     * @param _log Logging action status.
+     * @param _snackbar Display info about loading from backend for ldap group.
+     * @param _translate Translate messages for ldap group.
+     */
+    constructor(private _resources: LdapGroupResourceService,
+                private _log: LoggerService,
+                private _snackbar: SnackBarService,
+                private _translate: TranslateService) {
+        this._loading$ = new LoadingEmitter();
+        this._updateProgress$ = new LoadingEmitter();
+        this._ldapGroupsReload$ = new Subject<void>();
+        this._nextPage$ = new BehaviorSubject<number>(null);
+        this._endOfData = false;
+        this._pagination = {
+            size: 20,
+            totalElements: undefined,
+            totalPages: undefined,
+            number: -1
+        };
+        this._searchQuery = '';
+        this._clear = false;
+        this.negativeRolesQuery = new Array<string>();
+        this.rolesQuery = new Array<string>();
+
+        const ldapGroupsMap = this._nextPage$.pipe(
+            mergeMap(p => this.loadPage(p)),
+            tap(() => {
+                if (!this._clear) {
+                    this._ldapGroupsReload$.next();
+                }
+            }),
+            scan((acc, value) => {
+                const result = this._clear ? {} : {...acc, ...value};
+                this._clear = false;
+                return result;
+            }, {})
+        );
+        this._ldapGroups$ = ldapGroupsMap.pipe(
+            map(v => Object.values(v) as Array<LdapGroupListItem>),
+        );
+    }
+
+    ngOnDestroy(): void {
+        this._loading$.complete();
+        this._updateProgress$.complete();
+        this._ldapGroupsReload$.complete();
+        this._nextPage$.complete();
+    }
+
+    public get loading(): boolean {
+        return this._loading$.isActive;
+    }
+
+    public get loading$(): Observable<boolean> {
+        return this._loading$.asObservable();
+    }
+
+    public get ldapGroupsReload$(): Observable<void> {
+        return this._ldapGroupsReload$.asObservable();
+    }
+
+    public get ldapGroups$(): Observable<Array<LdapGroupListItem>> {
+        return this._ldapGroups$;
+    }
+
+    public get updating$(): Observable<boolean> {
+        return this._updateProgress$.asObservable();
+    }
+
+    public get totalLdapGroups(): number {
+        return this._pagination.totalElements ? this._pagination.totalElements : 0;
+    }
+
+    /**
+     * Get all ldapGroups from backend and mapped to [ldapGroupValue]{@link ldapGroupValue} interface with catching errors.
+     * @param page Page number that is requested. / Next page ldapGroups list.
+     */
+    public loadPage(page: number): Observable<{ [k: string]: LdapGroupListItem }> {
```

Comment:
can virtual scroll related methods be extracted to one implementation, or are they too dissimilar for such undertaking? I have a feeling that since we use virtual scrolling in multiple components, the driving stream implementation contains some bugfixes/QoL changes only in some places and not in others...

this would probably be best done in a separate Issue.

Reply:
create a new Issue :)